### PR TITLE
[FLINK-26460][table-planner]Fix Unsupported type when convertTypeToSpec: MAP

### PR DESCRIPTION
**Pull request by Tartarus0zm**
## What is the purpose of the change

Fix Unsupported type when convertTypeToSpec: MAP

## Brief change log

  - *PreValidateReWriter#maybeCast Add special handling for cast (null as MAP<>)*

## Verifying this change

* TableSinkTest#testInsertPartColumn
* TableSinkITCase#testInsertPartColumn

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by Tartarus0zm** [comment]
@wuchong  please take a look, if you have time, thanks

**Comment by Tartarus0zm** [comment]
@godfreyhe  please take a look, if you have time, thanks

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/PreValidateReWriter.scala [hunk 1]
```diff
@@ -391,8 +392,19 @@ object PreValidateReWriter {
       == desiredType)) {
       node
     } else {
-      SqlStdOperatorTable.CAST.createCall(SqlParserPos.ZERO,
-        node, SqlTypeUtil.convertTypeToSpec(desiredType))
+      // See FLINK-26460 for more details
+      val sqlDataTypeSpec =
+        if (currentType.getSqlTypeName == SqlTypeName.NULL
+          && desiredType.getSqlTypeName == SqlTypeName.MAP) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/PreValidateReWriter.scala:398]: SqlTypeUtil.isNull(currentType) && SqlTypeUtil.isMap(desiredType)

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/PreValidateReWriter.scala [hunk 2]
```diff
@@ -391,8 +392,19 @@ object PreValidateReWriter {
...
+          new SqlDataTypeSpec(
+            new SqlMapTypeNameSpec(SqlTypeUtil.convertTypeToSpec(desiredType.getKeyType),
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/PreValidateReWriter.scala:400]: we should consider whether the key type and value type are not null, such as: MAP<STRING NOT NULL, BIGINT>, MAP<STRING NOT NULL, BIGINT NOT NULL>

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution, I left some comments

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/PreValidateReWriter.scala (hunk 2 above)
Comment (Tartarus0zm) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/PreValidateReWriter.scala]: Thanks for remind, I has fix it.

**Comment by Tartarus0zm** [comment]
CI failure doesn't seem to be related to this patch

**Review by godfreyhe — APPROVED** [review]
LGTM

**Pull request closed without merging**